CA-184556: improve efficiency handling many VBDs during VM lifecycle operations

mg12 opened the pull request.

simonjbeaumont commented:
A few comments here. Think probably best we have a sync-up and talk through some of it. Maybe some of the concerns are unfounded but we need to check.

mg12 commented:
I have updated the patches with all the comments from the code review, and split it into two parts, one for the cache and the other for the Parallel atom. This pull request is for the cache only, I'll create a new one for the second part.

jonludlam commented on xc/xenops_server_xen.ml:
```diff
@@ -448,24 +448,87 @@ end
+	end
+	include Hashtbl
+	let create n = (create n, Mutex.create())
+	let discard (cache, mutex) domid = Mutex.execute mutex (fun () ->
```
Looks like there's some duplication going on here (2 include Hashtbls, 2 'let create' functions)

mg12 replied:
there are 2 levels of hashtables:
- the lowest level is a hashtable that keeps the cached devices for a specific domid
- the highest level is a hashtable the keeps the lowest level hashtables

The create for each of the levels has been redefined to return not only what create would return, but also a unique mutex for that hashtable instance.

simonjbeaumont commented:
OK, I think we've converged on this one.

simonjbeaumont commented:
Adding blocked label until review of #250 is finished
